core/src/main/java/org/elasticsearch/cluster/routing/IndexShardRoutingTable.java

```diff
@@ -312,6 +316,32 @@ public ShardIterator primaryFirstActiveInitializingShardsIt() {
         return new PlainShardIterator(shardId, ordered);
     }
 
+    public ShardIterator replicaActiveInitializingShardIt() {
+        if (!replicas.isEmpty() && !replicas.get(0).active() && !replicas.get(0).initializing()) {
+            List<ShardRouting> replicaList = ImmutableList.of();
+            return new PlainShardIterator(shardId, replicaList);
+        }
+        return replicaShardIt();
+    }
+
+    public ShardIterator replicaFirstActiveInitializingShardsIt() {
+        ArrayList<ShardRouting> ordered = new ArrayList<>(activeShards.size() + allInitializingShards.size());
+        // fill it in a randomized fashion
+        for (ShardRouting shardRouting : shuffler.shuffle(activeShards)) {
+            ordered.add(shardRouting);
+            if (shardRouting.primary() == false) {
+                // switch, its the matching node id
+                ordered.set(ordered.size() - 1, ordered.get(0));
+                ordered.set(0, shardRouting);
+            }
+        }
+        // no need to worry about replica first here..., its temporal
```

Comment:
we know there is only one primary, so we can optimize this case and push it to the end once we are done?